Introduce timescale calculations for VerticallyStretched grids

tomchor opened the pull request:
Closes https://github.com/CliMA/Oceananigans.jl/issues/1465

glwagner reviewed (commented):
Can we add a test?

Also, can we specialize the method for `cell_advection_timescale` to `RegularRectilinearGrid` so that the same bug doesn't occur for `RegularLatitudeLongitudeGrid`?

tomchor commented:
I just created this quickly so that I could pull the created branch locally and start coding. I think the final approach will be very different. I now think the best way is to especially `min_Δx`, etc, and use that everywhere we can. Although, like you said, we should probably consider calculating the CFL as a reduced field in the future.

tomchor commented:
Decide to leave `cell_advection_timescale` alone as it is. Instead I moved functions `grid_Δx` from `TurbulenceClosures` to `Grids/grid_timescales.jl` and specialized those functions to get the minimum of `Δz` for `VerticallyStretchedGrids`. This adds very few lines and also fixes another bug that happens when calculating the diffusive timescale for `VerticallyStretchedGrids`.

As soon as this merges into master I'll try to fix some of the bugs on https://github.com/CliMA/Oceananigans.jl/issues/1466

glwagner commented on src/Grids/grid_timescales.jl:
```diff
@@ -0,0 +1,18 @@
+
+min_Δx(grid) = grid.Δx
+min_Δy(grid) = grid.Δy
+min_Δz(grid) = grid.Δz
```
These should dispatch on `RegularRectilinearGrid` since they won't work for any other grid type.

glwagner replied:
In addition, it may be better to move these to `regular_rectilinear_grid.jl`.

glwagner commented on src/Grids/grid_timescales.jl:
```diff
@@ -0,0 +1,18 @@
+
+min_Δx(grid::VerticallyStretchedRectilinearGrid) = grid.Δx
+min_Δy(grid::VerticallyStretchedRectilinearGrid) = grid.Δy
+min_Δz(grid::VerticallyStretchedRectilinearGrid) = minimum(grid.Δzᵃᵃᶜ[1:grid.Nz])
```
It might be preferable to places these functions in `vertically_stretched_rectilinear_grid.jl`.

glwagner commented on src/Grids/Grids.jl:
```diff
@@ -107,5 +107,6 @@ include("input_validation.jl")
 include("regular_rectilinear_grid.jl")
 include("vertically_stretched_rectilinear_grid.jl")
 include("regular_latitude_longitude_grid.jl")
+include("grid_timescales.jl")
```
My comments below suggest moving this functionality into existing files. If we do keep this file, I think a more explicit name like `minimum_grid_spacings.jl` is better.

glwagner commented on src/Grids/grid_timescales.jl:
```diff
@@ -0,0 +1,18 @@
+min_Δz(grid::VerticallyStretchedRectilinearGrid) = minimum(grid.Δzᵃᵃᶜ[1:grid.Nz])
+
+
+
```
There are three extra lines here.

glwagner commented on src/Grids/grid_utils.jl:
```diff
@@ -300,3 +300,7 @@ function pop_flat_elements(tup, topo)
+
+
+
+
```
These four extra lines should be deleted.

glwagner commented on src/Utils/cell_advection_timescale.jl:
```diff
@@ -1,16 +1,19 @@
     return min(Δx/umax, Δy/vmax, Δz/wmax)
 end
 
+
```
This extra line should be deleted.

ali-ramadhan reviewed (approved):
Looks like a good temporary fix! 

But yeah might be good to add a test like this one but with a stretched grid: https://github.com/CliMA/Oceananigans.jl/blob/c34e6cd2166bbaa057186ffa795d348c1802485f/test/test_diagnostics.jl#L39-L51

Might also be useful for @mukund-gupta (X-Ref: #1429)

glwagner reviewed (changes requested):
Please add either a unit test for this function with `RegularRectilinearGrid` and `VerticallyStretchedRectilinearGrid`, or a test that `TimeStepWizard` works with `VerticallyStretchedRectilinearGrid`.

> Although, like you said, we should probably consider calculating the CFL as a reduced field in the future.

On #1422 we're discussing using `mapreduce` to avoid allocating memory for CFL calculations (and other purposes).

ali-ramadhan commented:
@tomchor Might also be good to tag a patch release (v0.53.1) so it's easier to make use of this fix.

tomchor commented:
Alright so as suggested I added a test for CFL calculation with a vertically stretched grid. Also refactored everything like @glwagner suggested, so if the tests pass I believe this should be good to go.

@ali-ramadhan should I just call JuliaRegistrator to register this after it's merged?

ali-ramadhan commented:
> @ali-ramadhan should I just call JuliaRegistrator to register this after it's merged?

Yes! Should work since you're a collaborator.

tomchor reviewed (commented):
@glwagner I don't know how to close your change request. I've already included a test and deleted all the stuff you told me to delete

glwagner commented:
>  I don't know how to close your change request.

If you "re-request a review", github sends me a notification to take another look!

glwagner commented on src/Grids/vertically_stretched_rectilinear_grid.jl:
```diff
@@ -206,3 +206,14 @@ Adapt.adapt_structure(to, grid::VerticallyStretchedRectilinearGrid{FT, TX, TY, T
+
+min_Δx(grid::VerticallyStretchedRectilinearGrid) = grid.Δx
+min_Δy(grid::VerticallyStretchedRectilinearGrid) = grid.Δy
+min_Δz(grid::VerticallyStretchedRectilinearGrid) = minimum(grid.Δzᵃᵃᶜ[1:grid.Nz])
```
You might need

```julia
min_Δz(grid::VerticallyStretchedRectilinearGrid) = minimum(view(grid.Δzᵃᵃᶜ, 1:grid.Nz))

tomchor replied:
Hmm, because of the GPU?

glwagner replied:
You also might be able to use

```julia
min_Δz(grid::VerticallyStretchedRectilinearGrid) = minimum(znodes(Center, grid))
```

glwagner replied:
Yes, I think you even want this on the CPU because `grid.Δzᵃᵃᶜ[1:grid.Nz]` allocates memory. We'll see if tests pass.

tomchor commented:
So, the test is failing on the GPU, but not on the CPU so I can't reproduce the error locally and I don't understand why it's failing. Can anyone give me a hand?

ali-ramadhan commented on test/test_diagnostics.jl:
```diff
@@ -12,32 +12,49 @@ function nan_checker_aborts_simulation(arch)
+TestModel_VerticallyStrectedRectGrid(arch, FT, ν=1.0, Δx=0.5) =
     IncompressibleModel(
-          grid = RegularRectilinearGrid(FT, size=(3, 3, 3), extent=(3Δx, 3Δx, 3Δx)),
+          grid = VerticallyStretchedRectilinearGrid(FT, size=(3, 3, 3), x=(0, 3Δx), y=(0, 3Δx), zF=0:Δx:3Δx,),
```
```suggestion
          grid = VerticallyStretchedRectilinearGrid(FT, architecture = arch, size=(3, 3, 3), x=(0, 3Δx), y=(0, 3Δx), zF=0:Δx:3Δx,),
```

ali-ramadhan replied:
@tomchor I accidently made a commit that I later realized was wrong so I reverted it.

I think the test was failing because the stretched grid needed a `architecture = arch` kwarg.

Maybe this is a reason to avoid having default architectures. Or at least we could check that the grid and model architectures are consistent as part of #1453.

tomchor replied:
That makes sense. Thanks!

tomchor commented:
@JuliaRegistrator register

JuliaRegistrator commented:
Comments on pull requests will not trigger Registrator, as it is disabled. Please try commenting on a commit or issue.
